library/coqlib.ml

```diff
@@ -16,29 +16,74 @@ open Libnames
 open Globnames
 open Nametab
 
-let coq = Libnames.coq_string (* "Coq" *)
+let make_dir l = DirPath.make (List.rev_map Id.of_string l)
 
 (************************************************************************)
-(* Generic functions to find Coq objects *)
+(* Coq reference API                                                    *)
+(************************************************************************)
+let coq = Libnames.coq_string (* "Coq" *)
 
-type message = string
+let init_dir = [ coq; "Init"]
 
-let make_dir l = DirPath.make (List.rev_map Id.of_string l)
+let jmeq_module_name       = [coq;"Logic";"JMeq"]
+let jmeq_module            = make_dir jmeq_module_name
 
 let find_reference locstr dir s =
   let dp = make_dir dir in
   let sp = Libnames.make_path dp (Id.of_string s) in
-  try Nametab.global_of_path sp
-  with Not_found ->
-    (* Following bug 5066 we are more permissive with the handling
-       of not found errors here *)
-    user_err ~hdr:locstr
-      Pp.(str "cannot find " ++ Libnames.pr_path sp ++
-          str "; maybe library " ++ DirPath.print dp ++
-          str " has to be required first.")
+  Nametab.global_of_path sp
 
 let coq_reference locstr dir s = find_reference locstr (coq::dir) s
 
+let table : (string, GlobRef.t Lazy.t) Hashtbl.t =
+  Hashtbl.create 127
+
+(** Can throw Not_found *)
+let lib_ref ?(quiet=false) s =
```

Comment:
I believe that using an option for the return type here is not a good idea. There are two use cases for this function:

 1. resolving a registered name; in which case getting `Not_found` is a bug and the printed warning helps to debug.
 2. checking whether some name is registered (e.g., to check if an optional dependency has been loaded), in which case the exception is not a bug (and `quiet` ought to be true).

Should there be two different functions for these two use cases?

I’ll improve the comments.

Reply:
In that case I'd rather raise an anomaly in the first case by default, and provide a `mem` function to check for the second one. Otherwise this is reenacting the terrible mess of `Anomaly: Uncaught exception Not_found` that is pervasive in our codebase due to map API raising exceptions at will.